### DISCUSSION TITLE
My experience with the program and hardware

### DISCUSSION THREAD

**Discussion by bp2008** [sipper69/Control3#7] [General]
Hello and thank you for making this!  I gave it a try and I think it works pretty well.

## CH340 keyboard/mouse cable

I used the keyboard/mouse emulator cable from your ebay listing: https://www.ebay.com/itm/1864468971043

I got CH340 drivers from the first Google hit, here: https://sparks.gogo.co.nz/ch340.html

## Video Capture Devices

I bought several capture devices from aliexpress since they are very cheap.

[Shape B](https://www.aliexpress.us/item/3256804804393973.html?spm=a2g0s.12269583.0.0.23004256adA229&gatewayAdapt=glo2usa), and [Shape C](https://www.aliexpress.us/item/3256804804393973.html?spm=a2g0s.12269583.0.0.23004256adA229&gatewayAdapt=glo2usa) from "Ciin Store"

![image](https://github.com/user-attachments/assets/e9052bbc-c59d-4236-ae46-007d5dbaa54e) ![image](https://github.com/user-attachments/assets/bca62b33-34e3-43f8-b2ad-c42bdd75d6bd)

The two pictured above from Ciin Store performed identically despite very different form factors, and had a noticeable amount of input lag in Control3.  For simple things like installing an OS or remote access software on a connected machine, it is fine.  I wouldn't want to use it a lot.  It is kind of like an older TV that isn't in game mode.

I also tried one from the "[Lemorele](https://www.aliexpress.us/item/3256806083704988.html?spm=a2g0o.order_list.order_list_main.11.4e8f1802ejdxbd&gatewayAdapt=glo2usa)" brand because it had a USB-C connector and was a little cheaper.  It has significantly **higher input lag** so I don't recommend it.
![image](https://github.com/user-attachments/assets/cfdc4072-4cc6-420d-be93-9b84a9c404f8)

## Control3 software wish list

This is very usable as it is, but a few things for the wish list would be:

* Making it not crash when a device it is using is unplugged.
* Add a menu to the video window:
  * It could let you change video capture modes without closing and reopening the video window (via a dropdown of video modes like the one in the main settings panel).
  * It could have a button to reset the size of the video window so the video renders at its native resolution (perhaps tricky since the host device could be using any amount of DPI scaling).
* The current video display resolution could be printed briefly as an overlay in one corner when the window is resized.
* End-to-end latency measurement tool built-in.
* Option to automatically open video window when running the app.
* Some way to send CTRL + ALT + DELETE to the connected device (it currently gets captured by the host).  Some apps have handled this by listening for CTRL + ALT + END and sending CTRL + ALT + DELETE instead to the remote device.
* When opening the video window, it should be sized such that the video player is at its native resolution (currently it is too small for 1080p).
* Render a crosshair or something where the HOST's cursor (mouse pointer) is, to reduce the perception of input lag when only seeing the remote cursor.

**Comment by sipper69** [comment]
Thank you for the great review! 

The lag is a result of two areas of delay.
1. Catching the mouse and key events, sending to and processing by the ch9329 via the serial port. 
2. Encoding the video by the MS2130 and let windows/app process and present the stream

So milliseconds delay and lag experience is a fact. 
Being acceptable is a personal experience as well, so I'm really curious what how big your lag is. Do we have an issue? 

My experience is perfectly usable for office use (no gaming of course). Can you check the VGA test I put on YouTube? That device is significantly worse than the MS2130, but still very usable. At least from my experience. You can check the lag on the double mouse pointer I show there. How do you experience this? 

Thank you for the wishlist. 

Yes graceful handling of unplugging is a top one. It should not crash though, I simply close the app. Lazy approach...

Btw, did you test with the latest MS App store version? Or this GitHub version? 

I'll probably can/will not spend time on further improvement. It was more a PoC/hobby project when I was in between jobs. But who knows...
Maybe someone in the community can come up with some genius work :-)

Cheers!

**Comment by sipper69** [comment]
Btw. i tested with the same, worked really great. 
But the YUV stream is big. Use MJPG.
Can be related?

**Comment by bp2008** [comment]
The "Lemorele" capture device had NV12 and MJPG options.  Both seemed to perform similarly in terms of latency.  The capture devices I got from "Ciin Store" only have YUV2 modes and their input lag was noticeably better but still over 100ms I would say.

Here are some screen recordings where I used 30 FPS rate, although 60 FPS should be technically better for latency, I have no convenient way to measure that.  (software wishlist: end-to-end latency measurement; of course that would require image analytics)

Lemorele: https://youtu.be/m22Frheofw0

Ciin: https://youtu.be/vuDwVGzs9PY

**Reply by sipper69** [reply]
Great! 
Yeah, your performance is the same as mine.
But that probably confirms how I interpret performance... But I'm sure you have no technical or app issues. This is the difference between our experiences, I'm an old office guy, you're a gamer?

Analyse latency is a cool function, that is something I'll do for sure (don't know when though).

Thanks man!

**Comment by bp2008** [comment]
I did see your "VGA test" on youtube.  Its latency looks to be very similar to what I get with the "Ciin" capture devices in YUV mode.  Are you saying you have a different HDMI capure device that has significantly better latency than that?

**Reply by sipper69** [reply]
Yeah, my ugreen with mjpg and 60 fps is really great. 

But I think my bar is a bit lower than yours :-)

**Comment by bp2008** [comment]
Also notable: for some reason your mouse pointer does not disappear when mousing over the video window (probably you were running an older version of Control3?), and on mine there are black bars to the right and left of the video which cause the mouse position to be offset improperly.

**Reply by sipper69** [reply]
I used a bug :-) 
But I have to check what (switching from full screen back?).

**Reply by sipper69** [reply]
Oww, now I see your comment. Improper offset because of black bars? Is related to window size, but that cannot be the reason. Never experienced a problem with mouse positioning. I have to figure out how to reproduce this.

**Reply by bp2008** [reply]
Yup.  Don't worry about the black bars.  It was a problem between the capture device and the client device (a laptop) sending the video.  The bars were part of the video feed from the capture device.  I got into Nvidia control panel on the laptop and changed its scaling method to "No scaling" and the black bars went away:

![image](https://github.com/user-attachments/assets/ed0dcb85-e5e8-472f-8f65-c8291489a847)

**Reply by bp2008** [reply]
And the black bars did not come back when I returned to "Aspect ratio" scaling mode.

**Comment by bp2008** [comment]
I have updated the software wish list above with several new items :)

**Reply by sipper69** [reply]
Thanks! 

Ctrl-alt-"d" does already cover one ;-)

**Reply by bp2008** [reply]
It sure does!!

**Reply by bp2008** [reply]
I see that mentioned in a tooltip now!
